text/0000-module-resolution.md

```diff
@@ -0,0 +1,356 @@
+
+### Why Objects instead of raw strings?
+
+In previous draft versions, `module` entries used to be just *strings* which had slightly better ergonomics. However, raw strings present a lot of ambiguities, let's look at one example:
+
+```json
+{
+    "modules": [
+        "just-string"
+    ]
+}
+```
+
+On the example below, it will be impossible to know if `just-string` is the name of an npm package, or a name of a folder that contains components, hence creationg resolution ambiguity.
+
+We considered other alternatives such as forcing specific prefix like `npm:package-name` or require a final slash (ex. `folder-name/`) for folders, but we believed that those were more cumbersome to learn and future-proof, so standarizing on an consistent object shape as ModuleRecord was a preferable and superior approach.
+
+### Types of ModuleRecord
+
+#### AliasModuleRecord
+
+This type of resolution maps a given module specifier to a particular path within the root of that module.
+
+```json
+{
+    "modules": [
+       {
+            "name": "ui/button",
+            "path": "src/modules/ui/button/button.js"
+        }        
+    ]
+}
+```
+
+`AliasModuleRecord` must contain `name` and a `path` keys.
+
+#### DirectoryModuleRecord
+
+This type of resolution allows to specify a folder that contain LWC modules. The structure of the folder its very specific (LWC has a very opinionated module structure). It must contain a namespace folder, then a folder per module bundle (moduleName) and then a moduleEntry file that matches the name of the bundle (the entry must be of type .css, .html, .js or .ts)
+
+::: note
+  The folder structure is something that the core LWC teams wants to enforce as a convention. There is no technical limitation on the compiler or other tools to resolve different folder structures. However we do want to provide and enforce as much as possible a canonical way so we can in the future keep iterating, improving and expanding the ergonomics of it.
+:::
+
+
+```json
+{
+    "modules": [
+       { "dir" : "src/modules" }
+    ]
+}
+```
+
+Here is an example of a valid module resolution given the previous DirectoryModuleRecord
+```
+src
+└── modules/ 
+    └── ui/ (namespace)
+        └── button/ (moduleName)
+            └── button.js (moduleEntry)
+```
+
+`DirectoryModuleRecord` must contain `dir` key.
+
+#### NpmModuleRecord
+
+This type of resolution tells the resolver to find an npm package with that name and resolve the modules it might have inside. This process can be recursive (see example at the top).
+
+```json
+{
+    "modules": [
+       { "npm" : "@ui/components" }     
+    ]
+}
+```
+
+`NpmModuleRecord` must contain `npm` key.
+
+## Algorithm
+
+Here is the algoritm for module resolution (the instructions are not in a rigurous specification format for brevity and time)
+
+```
+  Let `rootDir` be the initial path for the application, `initModules` 
```

Comment:
I was thinking more about this algorithm this morning and I don't think the original constraints forcing us to return a list of `ModuleRecordEntry` still holds with this proposal. 

Previously we needed to resolve ALL the module in advance because it was **too expensive** to search through the `node_modules` directory to find out what `package.json` has a `lwc` property. In order to get arround this issue, we exposed this `resolveModules` method that should ideally run once (before all the test, when starting a new compilation, ...). However this approach suffers from the fact that if you add a new LWC module when the application is running, you will have to restart the application for it to be consumed. The proposed algorithm also suffers from the same issue.

Now that each module have an explicit dependency mapping, there is a smaller set of possibility to resolve a given LWC module name. IMO we should deprecate the `resolveModules` method and expose a new `resolveModule` method that give a module name (`foo/bar`) and an importer (`./src/modules/ui/button/button.html`) should return a module entry point (`./nodes_modules/my-pkg/src/src/modules/foo/bar/bar.js`). This approach also aligns better with the [Node.js module resolution](https://nodejs.org/api/modules.html#modules_all_together).

Reply:
This algorithm still suffers from the same issue. For performance reasons, I don't think we should eagerly resolve all the modules in advance.